File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Control/Controller.java
package org.firstinspires.ftc.teamcode.Control;

import com.qualcomm.robotcore.hardware.Gamepad;

public class Controller extends Gamepad implements Runnable{

    private Gamepad gamepad;
    private boolean close = false;
    private boolean
            aPrev = false, bPrev = false, xPrev = false, yPrev = false,
            leftBumperPrev = false, rightBumperPrev = false, rightTriggerPrev = false, leftTriggerPrev = false;

    public Controller(Gamepad g){
        this.gamepad = g;
    }
    public boolean a() {return gamepad.a;}
    public boolean x() {return gamepad.x;}
    public boolean y() {return gamepad.y;}
    public boolean b() {return gamepad.b;}

    public boolean aOnPress() {return a() && !aPrev;}
    public boolean bOnPress() {return b() && !bPrev;}
    public boolean yOnPress() {return y() && !yPrev;}
    public boolean xOnPress() {return x() && !xPrev;}

    public float leftStickX(){return gamepad.left_stick_x;}
    public float leftStickY() {
        return gamepad.left_stick_y;
    }
    public float rightStickX() {
        return gamepad.right_stick_x;
    }
    public float rightStickY() {
        return gamepad.right_stick_y;
    }

    public boolean dPadUp() {return gamepad.dpad_up;}
    public boolean dPadDown() {
        return gamepad.dpad_down;
    }
    public boolean dPadLeft() {
        return gamepad.dpad_left;
    }
    public boolean dPadRight() {
        return gamepad.dpad_right;
    }

    public boolean leftBumper() {
        return gamepad.left_bumper;
    }
    public boolean rightBumper() {
        return gamepad.right_bumper;
    }

    public boolean leftBumperOnPress () {
        return leftBumper() && !leftBumperPrev;
    }
    public boolean rightBumperOnPress () {return rightBumper() && !rightBumperPrev;}

    public boolean leftStickButton() {
        return gamepad.left_stick_button;
    }
    public boolean rightStickButton() {
        return gamepad.right_stick_button;
    }

    public boolean leftTriggerPressed() {
        return leftTrigger() > 0;
    }
    public boolean rightTriggerPressed() {
        return rightTrigger() > 0;
    }
    public boolean leftTriggerOnPress() {
        return leftTriggerPressed() && !leftTriggerPrev;
    }
    public boolean rightTriggerOnPress() {return rightTriggerPressed() && !rightTriggerPrev;}

    public double getLeftStickAngle () {
        return Math.toDegrees(Math.atan2(leftStickY(), -leftStickX()));
    }
    public double getRightStickAngle () {
        return Math.toDegrees(Math.atan2(rightStickY(), -rightStickX()));
    }

    public double getTan (double x, double y) {
        double tan = -y/x;
        if (tan == Double.NEGATIVE_INFINITY || tan == Double.POSITIVE_INFINITY || tan != tan) tan = 0;
        return tan;
    }

    public float leftTrigger() {return gamepad.left_trigger;}
    public float rightTrigger() {return gamepad.right_trigger;}

    public boolean start() {return gamepad.start;}

    public synchronized void update(){
        aPrev = gamepad.a;
        bPrev = gamepad.b;
        xPrev = gamepad.x;
        yPrev = gamepad.y;
        leftBumperPrev = gamepad.left_bumper;
        rightBumperPrev = gamepad.right_bumper;
        rightTriggerPrev = rightTriggerPressed();
        leftTriggerPrev = leftTriggerPressed();
    }

    @Override
    public void run() {
        boolean close = false;
        while (!close) {
            System.out.println(aPrev);
            System.out.println(a());
            update();
            close = this.close;
            sleep(100);
        }
    }

    private static void sleep (int sleep) {
        try {Thread.sleep(sleep);}
        catch (InterruptedException e) {e.printStackTrace();}
    }

    public void close() {close = true;}
    public void startUpdate (){new Thread(this).start();}
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Control/Controller.kt
package org.firstinspires.ftc.teamcode.Control

import com.qualcomm.robotcore.hardware.Gamepad

class Controller(private val gamepad: Gamepad) : Gamepad(), Runnable {
    private var close = false
    private var aPrev = false
    private var bPrev = false
    private var xPrev = false
    private var yPrev = false
    private var leftBumperPrev = false
    private var rightBumperPrev = false
    private var rightTriggerPrev = false
    private var leftTriggerPrev = false

    val leftStickAngle: Double
        get() = Math.toDegrees(Math.atan2(leftStickY().toDouble(), (-leftStickX()).toDouble()))
    val rightStickAngle: Double
        get() = Math.toDegrees(Math.atan2(rightStickY().toDouble(), (-rightStickX()).toDouble()))

    fun a(): Boolean = gamepad.a
    fun x(): Boolean = gamepad.x
    fun y(): Boolean = gamepad.y
    fun b(): Boolean = gamepad.b

    fun aOnPress(): Boolean = a() && !aPrev
    fun bOnPress(): Boolean = b() && !bPrev
    fun yOnPress(): Boolean = y() && !yPrev
    fun xOnPress(): Boolean = x() && !xPrev

    fun leftStickX() = gamepad.left_stick_x
    fun leftStickY() = gamepad.left_stick_y
    fun rightStickX() = gamepad.right_stick_x
    fun rightStickY() = gamepad.right_stick_y

    fun dPadUp() = gamepad.dpad_up

    fun dPadDown(): Boolean {
        return gamepad.dpad_down
    }

    fun dPadLeft(): Boolean {
        return gamepad.dpad_left
    }

    fun dPadRight(): Boolean {
        return gamepad.dpad_right
    }

    fun leftBumper(): Boolean {
        return gamepad.left_bumper
    }

    fun rightBumper(): Boolean {
        return gamepad.right_bumper
    }

    fun leftBumperOnPress(): Boolean {
        return leftBumper() && !leftBumperPrev
    }

    fun rightBumperOnPress(): Boolean {
        return rightBumper() && !rightBumperPrev
    }

    fun leftStickButton(): Boolean {
        return gamepad.left_stick_button
    }

    fun rightStickButton(): Boolean {
        return gamepad.right_stick_button
    }

    fun leftTriggerPressed(): Boolean {
        return leftTrigger() > 0
    }

    fun rightTriggerPressed(): Boolean {
        return rightTrigger() > 0
    }

    fun leftTriggerOnPress(): Boolean {
        return leftTriggerPressed() && !leftTriggerPrev
    }

    fun rightTriggerOnPress(): Boolean {
        return rightTriggerPressed() && !rightTriggerPrev
    }

    fun getTan(x: Double, y: Double): Double {
        var tan = -y / x
        if (tan == java.lang.Double.NEGATIVE_INFINITY || tan == java.lang.Double.POSITIVE_INFINITY || tan != tan) tan = 0.0
        return tan
    }

    fun leftTrigger(): Float {
        return gamepad.left_trigger
    }

    fun rightTrigger(): Float {
        return gamepad.right_trigger
    }

    fun start(): Boolean {
        return gamepad.start
    }

    @Synchronized
    fun update() {
        aPrev = gamepad.a
        bPrev = gamepad.b
        xPrev = gamepad.x
        yPrev = gamepad.y
        leftBumperPrev = gamepad.left_bumper
        rightBumperPrev = gamepad.right_bumper
        rightTriggerPrev = rightTriggerPressed()
        leftTriggerPrev = leftTriggerPressed()
    }

    override fun run() {
        var close = false
        while (!close) {
            println(aPrev)
            println(a())
            update()
            close = this.close
            sleep(100)
        }
    }

    private fun sleep(sleep: Int) {
        try {
            Thread.sleep(sleep.toLong())
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

    }

    fun close() {
        close = true
    }

    fun startUpdate() {
        Thread(this).start()
    }
}

